nifi-nar-bundles/nifi-amqp-bundle/nifi-amqp-processors/src/main/java/org/apache/nifi/amqp/processors/AbstractAMQPProcessor.java

```diff
@@ -0,0 +1,204 @@
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.util.StandardValidators;
+
+import com.rabbitmq.client.Connection;
+import com.rabbitmq.client.ConnectionFactory;
+
+/**
+ * Base processor that uses RabbitMQ client API
+ * (https://www.rabbitmq.com/api-guide.html) to rendezvous with AMQP-based
+ * messaging systems version 0.9.1
+ *
+ * @param <T> the type of {@link AMQPWorker}. Please see {@link AMQPPublisher}
+ *            and {@link AMQPConsumer}
+ */
+abstract class AbstractAMQPProcessor<T extends AMQPWorker> extends AbstractProcessor {
+
+    public static final PropertyDescriptor HOST = new PropertyDescriptor.Builder()
+            .name("Host Name")
+            .description("Network address of AMQP broker (e.g., localhost)")
+            .required(true)
+            .defaultValue("localhost")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+    public static final PropertyDescriptor PORT = new PropertyDescriptor.Builder()
+            .name("Port")
+            .description("Numeric value identifying Port of AMQP broker (e.g., 5671)")
+            .required(true)
+            .defaultValue("5672")
+            .addValidator(StandardValidators.PORT_VALIDATOR)
+            .build();
+    public static final PropertyDescriptor V_HOST = new PropertyDescriptor.Builder()
+            .name("Virtual Host")
+            .description("Virtual Host name which segregates AMQP system for enhanced security.")
+            .required(false)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+    public static final PropertyDescriptor USER = new PropertyDescriptor.Builder()
+            .name("User Name")
+            .description("User Name used for authentication and authorization.")
+            .required(true)
+            .defaultValue("guest")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+    public static final PropertyDescriptor PASSWORD = new PropertyDescriptor.Builder()
+            .name("Password")
+            .description("Password used for authentication and authorization.")
+            .required(true)
+            .defaultValue("guest")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .sensitive(true)
+            .build();
+    public static final PropertyDescriptor AMQP_VERSION = new PropertyDescriptor.Builder()
+            .name("AMQP Version")
+            .description("AMQP Version. Currently only supports AMQP v0.9.1.")
+            .required(true)
+            .allowableValues("0.9.1")
+            .defaultValue("0.9.1")
+            .build();
+
+    static List<PropertyDescriptor> descriptors = new ArrayList<>();
+
+    /*
+     * Will ensure that list of PropertyDescriptors is build only once, since
+     * all other lifecycle methods are invoked multiple times
+     */
+    static {
```

Comment:
It took me a while to figure out what was going on in this block. I find this far more confusing than just explicitly stating the PropertyDescriptors to include.